hathor/verification/verification_model.py

```diff
@@ -0,0 +1,125 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from dataclasses import dataclass
+from typing import Generic, TypeAlias, TypeVar
+
+from typing_extensions import Self
+
+from hathor.daa import DifficultyAdjustmentAlgorithm
+from hathor.transaction import BaseTransaction, Block, MergeMinedBlock, TransactionMetadata, TxInput
+from hathor.transaction.exceptions import IncorrectParents, InexistentInput
+from hathor.transaction.storage.exceptions import TransactionDoesNotExist
+from hathor.transaction.token_creation_tx import TokenCreationTransaction
+from hathor.transaction.transaction import RewardLockedInfo, TokenInfo, Transaction
+from hathor.types import TokenUid
+from hathor.util import not_none
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class VertexDependencies:
```

Comment:
I think we can simplify it a lot by just implementing a simple memory database.

```python
class SimpleMemoryDB:
    def __init__(self):
        self._db: Dict[VertexId, BaseTransaction] = {}

    def add_vertex(self, vertex: BaseTransaction) -> None:
        self._db[vertex.hash] = vertex

    def get_vertex(self, _id: VertexId) -> BaseTransaction:
        return self._db[_id]
```

It might have support for metadata too if needed.

Reply:
This PR was completely changed, so I recommend re-reviewing it. Indeed the `SimpleMemoryDB` was a good solution for tx parents and spent txs, and I implemented it. But I needed to keep the verification model abstractions for some other dependencies.